File: ruby/custom-set/custom_set.rb
# OK, this passes all the tests, but c'mon...
#
# require 'set'
# class CustomSet < Set; end

class CustomSet
  def initialize(elements)
    @set = elements.zip([nil] * elements.length).to_h
  end

  def elements
    @set.keys.sort
  end

  def size
    @set.size
  end

  # methods that return a boolean

  def empty?
    @set.empty?
  end

  def member?(elem)
    @set.key? elem
  end

  def subset?(other)
    elements.none? { |e| !other.member?(e) }
  end

  def disjoint?(other)
    elements.none? { |e| other.member?(e) }
  end

  def ==(other)
    size == other.size && subset?(other)
  end

  # methods that return a CustomSet

  def add(elem)
    @set[elem] = nil
    self
  end

  def intersection(other)
    self.class.new(elements.select { |e| other.member?(e) })
  end

  def difference(other)
    self.class.new(elements.reject { |e| other.member?(e) })
  end

  def union(other)
    self.class.new(elements.concat(other.elements))
  end
end

File: javascript/custom-set/custom-set.js
/* eslint no-underscore-dangle: ["error", { "allowAfterThis": true }] */
/* eslint-disable  space-in-parens */

/* so, yeah, kind of cheating to use a Set as the storage class */

export class CustomSet {
  constructor(elements) {
    this._elements = new Set(elements);
  }

  // Note, Set.prototype.values() returns an *iterator*.
  // Converted that into an array.
  get elements() {
    return [...this._elements.values()];
  }

  get size() {
    return this._elements.size;
  }

  empty() {
    return this.size === 0;
  }

  add(element) {
    this._elements.add(element);
    return this;
  }

  contains(element) {
    return this._elements.has(element);
  }

  // true if all my elements are contained in the other set
  subset(other) {
    return this.elements.every(elem => other.contains(elem));
  }

  // true if two sets share no elements
  disjoint(other) {
    return this.elements.every(elem => !other.contains(elem));
  }

  eql(other) {
    return this.size === other.size && this.subset(other);
  }

  // elements contained in both sets
  intersection(other) {
    return new CustomSet( this.elements.filter(e => other.contains(e)) );
  }

  // which of my elements are _not_ in other
  difference(other) {
    return new CustomSet( this.elements.filter(e => !other.contains(e)) );
  }

  // all the elements.
  union(other) {
    return new CustomSet( this.elements.concat( other.difference(this).elements ));
  }
}
